Feature/domain entities

AlexBud369 opened the pull request:
# Core Domain Entities

**User**  
- Authentication (email, password hash, roles)  
- Manages owned books and refresh tokens  
- Account control (block/unblock)  

**Book**  
- Stores book metadata (title, author, description)  
- Tracks ownership (created by user)  
- Supports cover images and tags  

**Tag**  
- Categorizes books (genre, language, etc)  
- Many-to-many relationship with books  

**RefreshToken**  
- Extends user sessions securely  
- Tracks expiration and revocation

AlexBud369 committed:
refactor: fix project dependencies between API, Infrastructure and Application

AlexBud369 committed:
feat(domain): add User entity with authentication and book management

- Implemented User entity with fields for id, username, email, password hash, role, and book collection
- Added validation for username, email, and password hash
- Included methods for updating user details, role, and blocking status
- Supported EF Core with private constructor and private setters

AlexBud369 committed:
feat(domain): add Book entity for book management

- Implemented Book entity with fields for id, title, author, description, cover image, and tags
- Added validation for title and author
- Included methods for updating details, managing tags, and checking deletion permissions
- Supported EF Core with private constructor and private setters

AlexBud369 committed:
feat(domain): add RefreshToken entity for authentication

- Implemented RefreshToken entity with fields for id, token, user, creation, expiration, and revocation
- Added constructor with validation for token value
- Included Revoke method for invalidating tokens
- Supported EF Core with private constructor

AlexBud369 committed:
feat(domain): add Tag entity for book categorization

- Implemented Tag entity with fields for id, name and book collection
- Added private constructor for EF Core support and encapsulation
- Included validation through private setter for TagName
- Added UpdateTagName method for controlled modification

budnikme commented on src/Domain/Entities/Book.cs:
```diff
@@ -0,0 +1,127 @@
+﻿namespace Domain.Entities
+{
```
Please use file scoped namespaces

budnikme commented on src/Domain/Entities/Book.cs:
```diff
@@ -0,0 +1,127 @@
+{
+    public class Book
+    {
+        private Guid bookId;
```
Please use auto-properties with initializers instead, like: 
```
public string Title { get; set; }
```

budnikme commented on src/Domain/Entities/Book.cs:
```diff
@@ -0,0 +1,127 @@
+            CoverImageUrl = coverImageUrl;
+        }
+
+        public void AddTag(Tag tag)
```
Implement that as extension method, please

budnikme commented on src/Domain/Entities/Book.cs:
```diff
@@ -0,0 +1,127 @@
+            return user.UserId == createdByUserId || user.Role == Role.Admin;
+        }
+
+        private void ValidateNotEmpty(string value, string paramName)
```
It would be better to do such validation on the application or api level

budnikme commented on src/Domain/Entities/User.cs:
```diff
@@ -0,0 +1,173 @@
+        }
+
+
+        public void UpdateName(string newName)
```
I guess that all these methods should be in the services

AlexBud369 committed:
fix: refactor Book entity
- replaced private fields with auto-properties
- removed methods UpdateDetails, UpdateCoverImageUrl, AddTag, RemoveTag, CanBeDeletedBy
-removed ValidateNotEmpty from constructor.

AlexBud369 committed:
fix: refactor User entity
- replaced private fields with auto-properties
- removed methods UpdateName, UpdateEmail, UpdatePasswordHash, ChangeRole, Block, Unblock, AddBook, AddRefreshToken
- removed ValidateNotEmpty from constructor
- used IReadOnlyList<T> for RefreshTokens and Books with private backing Lists
- fixed redundant UserId initialization in constructor

AlexBud369 committed:
fix: refactor Tag entity
- replaced private fields with auto-properties
- removed UpdateTagName method
- removed redundant tagId parameter from constructor
- used IReadOnlyList<Book> for Books with private backing List
- removed validation from constructor

AlexBud369 committed:
fix: refactor RefreshToken entity
- converted to file-scoped namespace
- replaced private fields with auto-properties
- remove validation in constructor
